src/utils/react-query-factory.js

```diff
@@ -0,0 +1,72 @@
+// The keys that map to enterpriseQueryKeys that is used to define specific queries
+// and used to specifically invalidate queries
+export const queryKeys = {
+  // Global key to all queries
+  all: ['enterprise'],
+  enterprise: (enterpriseUuid) => [...queryKeys.all, enterpriseUuid],
+  entitlements: ['user', 'entitlements'],
+  enterpriseLearner: () => [...queryKeys.all, 'linked-enterprise-customer-users'],
+  enterpriseCourseEnrollments: (enterpriseUuid) => [...queryKeys.enterprise(enterpriseUuid), 'enrollments'],
+  enterpriseCourseMetadata: (enterpriseUuid) => [...queryKeys.enterprise(enterpriseUuid), 'course'],
+  canRedeem: (enterpriseUuid) => [...queryKeys.enterpriseCourseMetadata(enterpriseUuid), 'can-redeem'],
```

Comment:
IMO, it's a bit odd for `canRedeem` to rely on `enterpriseCourseMetadata`, given `canRedeem` does not deal with course metadata at all.

Would it make more sense to have a standalone `course` group that both `enterpriseCourseMetadata` and `canRedeem` rely on?

FWIW, it does appear the @lukemorales/query-key-factory NPM package could help better implement these types of groupings, both from a readability and authoring POV, e.g.:

```jsx
const enterprise = createQueryKeys('enterprise', {
  enterpriseCustomer: (enterpriseId) => ({
    queryKey: [enterpriseId],
    contextQueries: {
      // ...
      course: (courseKey) => ({
        queryKey: [courseKey],
        contextQueries: {
          contentMetadata: {
            queryKey: null,
            queryFn: async ({ queryKey }) => fetchCourseMetadata(queryKey[2], queryKey[4]),
          },
          canRedeem: (availableCourseRunKeys) => ({
            queryKey: [availableCourseRunKeys],
            queryFn: async ({ queryKey }) => fetchCanRedeem(queryKey[2], availableCourseRunKeys),
          }),
        },
      }),
    },
  }),
});
```

The above would result in something like the following to be used in the route loaders and the related custom hooks rendering `useQuery`:

```jsx
export default function useCourseMetadata() {
  const { courseKey } = useParams();
  const { data: { enterpriseCustomer } } = useEnterpriseLearner();
  const enterpriseId = enterpriseCustomer.uuid;
  return useQuery(
    queries.enterprise.enterpriseCustomer(enterpriseId)._ctx.course(courseKey)._ctx.contentMetadata,
  );
}
```

... where the contextual child queries are accessed via `._ctx` (it will throw an ESLint error about dangling underscore, but we can disable the ESLint rule for `._ctx` specifically, if we opt to go this route).

Also, note in the above that, to your earlier point around having the opportunity to nix the `makeEnterpriseLearnerQuery`, etc. type functions by co-locating `queryFn` with the query key definitions. Curious to hear what you think about this?